### PR TITLE
Make the network API non Linux specific

### LINKED ISSUE #3005: Kata host networking implementation is Linux specific
As part of #3004, we would like to have Kata's sandbox host networking support non-Linux UNIXes.

At the moment, the `Network` Kata implementation is mostly based on Linux networking namespaces and a Kata sandbox network contains `Endpoint` that, for most of them, are Linux specific as well.

Instead we'd like a Kata sandbox `Network` to be defined as an interface and have separate implementations for each operating system that we'd like to see Kata run on top of.

### PR DESCRIPTION
**Pull request by sameo**
This fixes #3005 by removing all Linux specific bits from our Sandbox networking implementation.
In particular, it removes the dependencies to the Linux netns packages and convert the `Network` structure into an interface. That will allow us to define per host OSes implementation for managing host networking.

### CODE REVIEW

File: src/runtime/virtcontainers/sandbox.go
Comment (egernst) [inline src/runtime/virtcontainers/sandbox.go:808]: Remind me; we are setting the sandbox's network in both newSandbox and createNetwork -- are these executed in differing flows, or is this duplicated?

File: src/runtime/virtcontainers/network.go
Comment (egernst) [inline src/runtime/virtcontainers/network.go]: nit: s/Enpoint/Endpoint

File: src/runtime/virtcontainers/network.go
Comment (egernst) [inline src/runtime/virtcontainers/network.go]: Can you clarify this part?

File: src/runtime/virtcontainers/network.go
Comment (egernst) [inline src/runtime/virtcontainers/network.go]: Not for changing in this PR per se, but I think it'd be better if we had our own defined type for NetworkInfo in this package, and had persistapi use this (instead of the opposite and relying on that otherwise unrelated package). Would you have any concerns/input on this, if I were to change this in a follow on PR?

**Review by egernst — COMMENTED** [review]
13 commits of great cleanup. Left a couple of nits/questions. Generally looks good to me.

**Comment by egernst** [comment]
Saw this lint: 
```
Error: network.go:97:18: fieldalignment: struct with 480 pointer bytes could be 464 (govet)
795
``` 

I hate field alignment error reports with the passion of a thousand burning suns.  I am not sure these 26 Bytes of savings are worth it, particularly if we don't have tools for fixing automatically, and because often it sacrifices readability/organization of the struct. Opened https://github.com/kata-containers/kata-containers/issues/3604 to discuss.

File: src/runtime/virtcontainers/network.go
Comment (egernst) [inline src/runtime/virtcontainers/network.go:2]: nit: typo: should be Apple

File: src/runtime/virtcontainers/network.go
Comment (egernst) [inline src/runtime/virtcontainers/network.go:17]: It seems that you removed *some* of the netlink utilization here -- we still have netlink included for some of the structure definitions which still sit in this file?

File: src/runtime/virtcontainers/network.go
Comment (egernst) [inline src/runtime/virtcontainers/network.go:91]: Still dependent on netlink?

File: src/runtime/virtcontainers/sandbox.go
Comment (sameo) [reply src/runtime/virtcontainers/sandbox.go:808]: It's a dup, carried over from the switch between netns to the `Network` structure and then interface.
I pushed an additional commit to simplify `createNetwork()`.

File: src/runtime/virtcontainers/network.go
Comment (sameo) [reply src/runtime/virtcontainers/network.go]: Fixed.

File: src/runtime/virtcontainers/network.go
Comment (sameo) [reply src/runtime/virtcontainers/network.go:2]: Fixed.

File: src/runtime/virtcontainers/network.go
Comment (sameo) [reply src/runtime/virtcontainers/network.go]: I added a comment in the `network_linux.go` file for clarification:

```
// NewNetwork creates a new Linux Network from a NetworkConfig.
// The constructor is overloaded as it can be called with 0 or 1
// argument. The former is used to create empty networks, mostly
// for unit testing. Passing more than one NetworkConfig pointer
// will make the constructor fail.
```

**Comment by sameo** [comment]
> Saw this lint:
> 
> ```
> Error: network.go:97:18: fieldalignment: struct with 480 pointer bytes could be 464 (govet)
> 795
> ```
> 
> I hate field alignment error reports with the passion of a thousand burning suns. I am not sure these 26 Bytes of savings are worth it, particularly if we don't have tools for fixing automatically, and because often it sacrifices readability/organization of the struct. Opened #3604 to discuss.

FWIW, all linter warnings should be fixed now.

File: src/runtime/virtcontainers/network.go
Comment (sameo) [reply src/runtime/virtcontainers/network.go]: oops. Fixed as well.

**Review by egernst — DISMISSED** [review]

**Comment by fidencio** [comment]
Let me add a do-not-merge label as I really want to go through this (and other big PRs that were opened recently).

**Comment by jodh-intel** [comment]
I haven't reviewed this yet (hopefully Monday!) but for changes like this, we need to keep in mind the potential need to update the docs (and diagrams), in particular API docs and possibly the arch doc (https://github.com/kata-containers/kata-containers/blob/main/docs/design/architecture/networking.md).

**Review by bergwolf — APPROVED** [review]
Indeed nice cleanup! Thanks @sameo!

**Comment by fidencio** [comment]
@sameo, first of all, thanks for breaking up the patches in a way those are easy to review, very much appreciated.

I second James comments about updating the documentation and I'll leave it up to your discretion to either do this now, as part of this PR (which I think that would be the best possible case), or as a coming up PR.

Apart from this, lgtm!  I'd like to have this merged sooner than later to avoid conflicts that may come from smaller patches touching this area (I have at least one series that can wait till this one gets merged).

**Review by fidencio — DISMISSED** [review]

**Comment by fidencio** [comment]
(I've added back the `do-not-merge label` so @sameo can remove it depending on his decision to update the docs now or later).

**Comment by fgiudici** [comment]
Just for information: I'm a bit late, but I would like and will take a look at this by EoD (don't want to slow down merging, I will take a look also if it gets merged in the meanwhile).

**Comment by sameo** [comment]
> Just for information: I'm a bit late, but I would like and will take a look at this by EoD (don't want to slow down merging, I will take a look also if it gets merged in the meanwhile).

Thanks for the heads up @fgiudici . I'll wait for your review, take your time.

**Comment by sameo** [comment]
@jodh-intel @fidencio Thanks for the review and the reminder about the documentation update.

This PR does not carry any external API networking changes, but does change `NetworkConfig` (To make it not Linux netns bound) and so I now updated the virtcontainers `api.md` file.

**Comment by sameo** [comment]
> (I've added back the `do-not-merge` label so @sameo can remove it depending on his decision to update the docs now or later).

I updated the docs, but I'll wait for @fgiudici review before removing the label.

**Review by fgiudici — APPROVED** [review]
Many thanks @sameo  for waiting on me :-)
I was not that familiar with the code here, but this rework looks quite good to me!
Thanks

**Review by egernst — APPROVED** [review]

**Pull request merged**
